### DISCUSSION TITLE
Custom hooks with QueryOptions & TypeScript

### DISCUSSION THREAD

**Discussion by fnn** [TanStack/query#1195] [Q&A] [answered]
Hey everyone,

I would like to create custom hooks with the ability to pass the query options.

What I do have so far:
```typescript
const usePost = (id: number, options?: UseQueryOptions<Post>) =>
  useQuery(["post", id], getPost, options)
```

This is kind of working but the typing for the select function is wrong.

```typescript
const { data } = usePost(1234, { select: (post) => post.author });
```
Will lead to this error:
```
Type '(post: Post) => string' is not assignable to type '(data: Post) => Post'.
  Type 'string' is not assignable to type 'Post'.
```

I don't know how to set up the options types to get the same result as useQuery.
It would be awesome to have pretty much the useQuery interfaces just with predefined `[entity, id]` query key and `getEntity` fetch function.

**Comment by dusty** [comment]
Try passing the result set type to useQuery 

```
const usePost = (id: number, options?: UseQueryOptions<Post>) => {
  useQuery<Post>(["post", id], getPost, options)
}
```

https://github.com/tannerlinsley/react-query/blob/979527b2a922ece2c548f1a82de74052138368cb/src/react/useQuery.ts#L29

**Reply by fnn** [reply]
Thank you for your answer, but that does not work. The error is the same.

**Comment by fnn** [comment, marked as answer]
Okay, I'm really bad at TypeScript but this is what I did and I think it is working:
```typescript
const usePost = <T = Post>(id: number, options?: UseQueryOptions<T, any, Post>) =>
  useQuery(["post", id], getPost, options)
```
And now if I use the custom hook with a select function the data type is correct.
```typescript
const { data } = usePost(1234, { select: (post) => post.author }); // 'data' is a string
```

I just don't know why this works. I thought with this solution I always have to pass the type to the custom hook like this:
`usePost<string>(1234, { select: (post) => post.author });`, but this is not the case..

**Reply by TkDodo** [reply]
> I just don't know why this works. I thought with this solution I always have to pass the type to the custom hook like this:

you would need to do that, if it weren't for your default type parameter (`T = Post`). Just like default values in JavaScript, default values on type level are used when the generic is not passed or cannot be inferred.

Since `T` is inferred based on the return value of the `select` function, it can be nicely inferred, so if you make your select function mandatory, the default value on type level would not be needed. But since select is optional in the `UseQueryOptions`, the compiler will fall back to `Post` as a type for `T` if no select function was passed.

Since I don't usually expose all the QueryOptions if I create a custom hook, I made a version that only takes the `select` function as second parameter.

https://codesandbox.io/s/practical-noyce-42z0g

Two versions actually, one where select is optional and one where it is mandatory. You can see that the `T = Post` can be omitted in the mandatory version.
This also has the advantage of not having to use `UseQueryOptions` for typing things :)

**Reply by fnn** [reply]
> ...or cannot be inferred.

I think this is the important part. Default values in JavaScript don't have that. There it is either passed value or default value.

I like the flexibility of passing the options in my specific use case. If I have a lot of scenarios where I always need the same options I will build and abstraction around this. 
Thanks a lot for your comment and the sandbox, helped me a lot!

**Reply by TkDodo** [reply]
> Default values in JavaScript don't have that. There it is either passed value or default value.

yes, you are right about that 👍 

> I like the flexibility of passing the options in my specific use case. If I have a lot of scenarios where I always need the same options I will build and abstraction around this.

sure, I'm glad you found the solution for that scenario as well.

> Thanks a lot for your comment and the sandbox, helped me a lot!

no worries, happy to help.

**Reply by zakariaelas** [reply]
Hello @TkDodo , sorry to piggyback on this! 

I'm using react-query `^3.29.0`, and the codesandbox demo you provided breaks the first version (which i'm interested in) of the `usePosts` hook under`3.29.0`. 

Using the code in the demo, it seems that the query key type is being inferred to `(string | number)[]` conflicting with the default `QueryKey` type coming from the `UseQueryOptions<Post, unknown, Post>` (because of the default 4th generic type, this is the same as `UseQueryOptions<Post, unknown, Post, Query>`). 

How do you suggest solving this? An easy and quick fix is to do `UseQueryOptions<Post, unknown, Post, any[]>`. Would you advise against this? I could do `UseQueryOptions<Post, unknown, Post, (string | number)[]>` but I think is adds some friction and I'd rather infer types as much as possible. 

Thanks!

**Reply by TkDodo** [reply]
This has been reported multiple times, please see: https://github.com/tannerlinsley/react-query/issues/2795

**Reply by piotrpalek** [reply]
@TkDodo sorry to dig this up, but I'm trying to workaround the "wrap react-query and pass all custom options to it with typescript working" problem. Did anything change? I've looked at the comment you've linked, but the solution there also doesn't work when you try to use `select` right?

I mean when I type the `UseQueryOptions` with `Thing`, and will try to pass a `select` option through my wrapper it'll complain that it doesn't return `Thing`.

**Reply by TkDodo** [reply]
you need to use all 4 generics. I'm not really sure why you'd want to create a low-level abstraction that allows all options, but if you really need to, this twitter thread describes a possible approach:

https://twitter.com/TkDodo/status/1491451513264574501

**Reply by piotrpalek** [reply]
@TkDodo thanks so much! Clearly I need to figure out ReactQuery more if that's not a pattern that should really be followed (feels natural to me though), anyway, ty :)

**Reply by raky291** [reply]
@TkDodo Thanks for sharing the tweet!

Would it be possible to add that information to the documentation? I think it's very valuable.

**Comment by mathura333** [comment]
After allot of hit and trails finally I wrote this and it worked perfectly


```
type ApiBody = mapApiTypes['/api/v1/shortlist_cta/:id']['res']['body'];

const useShortlistCta = <QueryRes = ApiBody>(
  { id, by = 'merchant_id' }: ApiParams,
  select?: (data: ApiBody) => QueryRes,
  options: UseQueryOptions<ApiBody, unknown, QueryRes> = {}
) => {
  return useQuery<ApiBody, unknown, QueryRes>(
    [RQ_SHORTLIST_CTA, id, by],
    () => fetchShortlistCta(id, by),
    {
      select,
      enabled: !!id,
      ...options,
    }
  );
};

```

  ```
const { data } = useShortlistCta(
    {
      id: '123',
    },
    (resData) => resData[0].enabled
  ) 

// here data type is boolean | undefined as output which is as expected
```

**Comment by Yuqgnort** [comment]
Hello this is my solution for useQuery and useMutation

![image](https://github.com/TanStack/query/assets/103628596/2a30e404-fd42-4a12-a5ba-025cf2445c41)

![image](https://github.com/TanStack/query/assets/103628596/0b4b2cbd-70f9-4164-9363-8f2614617ca1)

**Reply by TkDodo** [reply]
have you tried using the `select` option with your abstraction? Because it won't work ... it would be a lot better to just use the `queryOptions` function from v5:

```
const getReeivingDepartmentQueryOptions = queryOptions({
  querykey: [...],
  queyFn: ...
})
```

then, on call sides, just do:

```
useQuery({
  ...getReeivingDepartmentQueryOptions,
  ...moreOptionsGoHere
})
```

**Reply by Yuqgnort** [reply]
amazing,it's very helpful,thanks bro

**Reply by shongs27** [reply]
@TkDodo So, then how to make `moreOptionsGoHere` type?
```
const getReeivingDepartmentQueryOptions = queryOptions({
  querykey: [...],
  queyFn: ...
})
```
```
export const useGetNormalActivityList = (
  date?: string,
  options?: Omit<UseQueryOptions<any>, 'queryKey' | 'queryFn'>, // <- here?
) => {
  return useQuery({
    ...getReeivingDepartmentQueryOptions,
    ...options,
  });
```

**Reply by TkDodo** [reply]
> So, then how to make moreOptionsGoHere type?

You just inline them and don't further abstract them to a custom hook - that's the whole point:

```
const getReeivingDepartmentQueryOptions = queryOptions({
  querykey: [...],
  queyFn: ...
})

// usage 1 in <MyComponent />

const { data } = useQuery({
 ...getReeivingDepartmentQueryOptions,
 enabled: !!someCondition,
}

// usage 2 in <MyOtherComponent />

const { data } = useSuspenseQuery({
 ...getReeivingDepartmentQueryOptions,
 refetchInterval: 10 * 1000,
}
```

**Reply by jknb** [reply]
> > So, then how to make moreOptionsGoHere type?
> 
> You just inline them and don't further abstract them to a custom hook - that's the whole point:
> 
> ```
> const getReeivingDepartmentQueryOptions = queryOptions({
>   querykey: [...],
>   queyFn: ...
> })
> 
> // usage 1 in <MyComponent />
> 
> const { data } = useQuery({
>  ...getReeivingDepartmentQueryOptions,
>  enabled: !!someCondition,
> }
> 
> // usage 2 in <MyOtherComponent />
> 
> const { data } = useSuspenseQuery({
>  ...getReeivingDepartmentQueryOptions,
>  refetchInterval: 10 * 1000,
> }
> ```

Are there no cases where making an abstraction makes sense? All the queries in my app need to do a common thing (call a hook to get an important state variable and use it in the fetch function. Also the fetch function is used everywhere). 

Repeating the same thing in every component feels too repetitive instead of creating a wrapper hook that does this. I want to make it generic so it can be used anywhere and any consumer can pass the options they want.

Maybe my approach is wrong and I'm looking at the wrong problem but the way I've implemented it, I'd like to be able to somehow pass any options as params and type the select properly.

**Reply by TkDodo** [reply]
> instead of creating a wrapper hook that does this

sure you can do that, but don’t make a wrapper hook that again takes arbitrary options because then it’s no longer “all the queries do a common thing”. If it’s a common thing, sure, make a hook:

```
const useMyQueryWithInterval = () => {
  return useSuspenseQuery({
   ...getReeivingDepartmentQueryOptions,
   refetchInterval: 10 * 1000,
  })
}
```

or make another query options abstraction:

```
const getReeivingDepartmentQueryOptionsWithInterval = () => {
  return queryOptions({
   ...getReeivingDepartmentQueryOptions,
   refetchInterval: 10 * 1000,
  })
}
```

but don’t try to do “useReeivingDepartmentQuery(arbitraryOptionsGoHere)” because that’s just what `queryOptions` or `useQuery` already do.
